openstreetmap-carto.lua

```diff
@@ -0,0 +1,518 @@
+        -- Avoid generating objects for untagged boundary relations
+        if next(keyvalues) ~= nil then
+            keyvalues.z_order = z_order(keyvalues)
+            return 0, keyvalues, members_superseeded, 1, 0, roads(keyvalues)
+        end
+    -- For multipolygons...
+    elseif (type == "multipolygon") then
+        -- Multipolygons by definition are polygons, so we know roads = linestring = 0, polygon = 1
+        -- The type tag has been removed, so this checks for untagged MPs
+        if next(keyvalues) == nil then
+            -- This is an old-style MP
+            local combined_tags = combine_member_tags(keyvaluemembers)
+            if combined_tags == nil then
+                -- This is an invalid old-style MP with conflicting tags
+                return 1, {}, members_superseeded, 0, 1, 0
+            elseif next(combined_tags) == nil then
+                -- This is a valid old-style MP, but has no tags
+                return 1, {}, members_superseeded, 0, 1, 0
+            else
+                -- This is a valid old-style MP because a set of tags could be
+                -- made. For each member, the POLYGON its way tags generated is
+                -- superseded by the geometry from the MP, or the way was untagged.
+                -- Untagged ways generate no geom, so can be superseded too.
+                for i = 1, membercount do
+                    members_superseeded[i] = 1
+                end
+                combined_tags.z_order = z_order(keyvalues)
+                return 0, combined_tags, members_superseeded, 0, 1, 0
+            end
+        else
+            -- This is a new-style MP
+            keyvalues.z_order = z_order(keyvalues)
+            return 0, keyvalues, members_superseeded, 0, 1, 0
+        end
+        assert(false, "End of control reached prematurely for MP")
+    elseif type == "route" then
+        if next(keyvalues) ~= nil then
+            keyvalues.z_order = z_order(keyvalues)
+            return 0, keyvalues, members_superseeded, 1, 0, roads(keyvalues)
+        end
+    end
+
+    -- Untagged or unknown type
+    return 1, keyvalues, members_superseeded, 0, 0, 0
+end
+
+--- Check if an object with given tags should be treated as polygon
+-- @param tags OSM tags
+-- @return 1 if area, 0 if linear
+function isarea (tags)
+    -- Treat objects tagged as area=yes polygon, other area as no
+    if tags["area"] then
+        return tags["area"] == "yes" and 1 or 0
+    end
+
+   -- Search through object's tags
+    for k, v in pairs(tags) do
+        -- Check if it has a polygon key and not a linestring override, or a polygon k=v
+        for _, ptag in ipairs(polygon_keys) do
+            if k == ptag and not (linestring_values[k] and linestring_values[k][v]) then
+                return 1
+            end
+        end
+
+        if (polygon_values[k] and polygon_values[k][v]) then
+            return 1
+        end
+    end
+    return 0
+end
+
+function is_in (needle, haystack)
```

Comment:
So it is. I remember @math1985 adding it for a reason, but maybe I removed that reason with the MP handling change. There's probably a couple places where it **should** be used.